src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProvider.java

```diff
@@ -47,72 +56,166 @@ public static StreamingClientProvider getStreamingClientProviderInstance() {
-      // recreate streaming client if needed
-      if (!StreamingClientHandler.isClientValid(this.parameterEnabledClient)) {
-        LOGGER.error("Current streaming client is invalid, recreating client");
-        this.parameterEnabledClient = this.streamingClientHandler.createClient(connectorConfig);
+          "Streaming client optimization is enabled per worker node. Reusing valid clients when"
+              + " possible");
+      resultClient = this.registeredClients.get(clientProperties);
+
+      // refresh if registered client is invalid
+      if (!StreamingClientHandler.isClientValid(resultClient)) {
+        resultClient = this.streamingClientHandler.createClient(clientProperties);
+        this.registeredClients.put(clientProperties, resultClient);
       }
-      this.providerLock.unlock();
-      return this.parameterEnabledClient;
     } else {
-      LOGGER.info("Streaming client optimization is disabled, creating a new streaming client");
-      return this.streamingClientHandler.createClient(connectorConfig);
+      resultClient = this.streamingClientHandler.createClient(clientProperties);
+      LOGGER.info(
+          "Streaming client optimization is disabled, creating a new streaming client with name:"
+              + " {}",
+          resultClient.getName());
     }
+
+    return resultClient;
   }
 
   /**
-   * Closes the given client
+   * Closes the given client and deregisters it from the cache if necessary. It will also call close
+   * on the registered client, which should be the same as the given client so the call will no-op.
    *
+   * @param connectorConfig The configuration to deregister from the cache
    * @param client The client to be closed
    */
-  public void closeClient(SnowflakeStreamingIngestClient client) {
-    this.providerLock.lock();
+  public void closeClient(
+      Map<String, String> connectorConfig, SnowflakeStreamingIngestClient client) {
+    StreamingClientProperties clientProperties = new StreamingClientProperties(connectorConfig);
+
+    // invalidate cache
+    SnowflakeStreamingIngestClient registeredClient =
+        this.registeredClients.getIfPresent(clientProperties);
+    if (registeredClient != null) {
+      // invalidations are processed on the next get or in the background, so we still need to close
+      // the client here
+      this.registeredClients.invalidate(clientProperties);
+      this.streamingClientHandler.closeClient(registeredClient);
+    }
+
+    // also close given client in case it is different from registered client. this should no-op if
+    // it is already closed
     this.streamingClientHandler.closeClient(client);
-    this.providerLock.unlock();
+  }
+
+  // TEST ONLY - return the current state of the registered clients
+  @VisibleForTesting
+  public Map<StreamingClientProperties, SnowflakeStreamingIngestClient> getRegisteredClients() {
+    return this.registeredClients.asMap();
+  }
+
+  public static class StreamingClientProperties {
```

Comment:
short description here would be useful! but I like the idea.

Reply:
Split into a separate class with comments, PTAL!